src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs

```diff
@@ -0,0 +1,235 @@
+    {
+        // Having a key modifier ensures with overwhelming probability that no two encryption operations
+        // will ever derive the same (encryption subkey, MAC subkey) pair. This limits an attacker's
+        // ability to mount a key-dependent chosen ciphertext attack. See also the class-level comment
+        // for how this is used to overcome GCM's IV limitations.
+        private const int KEY_MODIFIER_SIZE_IN_BYTES = 128 / 8;
+
+        private const int NONCE_SIZE_IN_BYTES = 96 / 8; // GCM has a fixed 96-bit IV
+        private const int TAG_SIZE_IN_BYTES = 128 / 8; // we're hardcoding a 128-bit authentication tag size
+
+        // 128 "00-01-00-00-00-10-00-00-00-0C-00-00-00-10-00-00-00-10-95-7C-50-FF-69-2E-38-8B-9A-D5-C7-68-9E-4B-9E-2B"
+        private static readonly byte[] AES_128_GCM_Header = new byte[] { 0x00, 0x01, 0x00, 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x0C, 0x00, 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x10, 0x95, 0x7C, 0x50, 0xFF, 0x69, 0x2E, 0x38, 0x8B, 0x9A, 0xD5, 0xC7, 0x68, 0x9E, 0x4B, 0x9E, 0x2B };
+
+        // 192 "00-01-00-00-00-18-00-00-00-0C-00-00-00-10-00-00-00-10-0D-AA-01-3A-95-0A-DA-2B-79-8F-5F-F2-72-FA-D3-63"
+        private static readonly byte[] AES_192_GCM_Header = new byte[] { 0x00, 0x01, 0x00, 0x00, 0x00, 0x18, 0x00, 0x00, 0x00, 0x0C, 0x00, 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x10, 0x0D, 0xAA, 0x01, 0x3A, 0x95, 0x0A, 0xDA, 0x2B, 0x79, 0x8F, 0x5F, 0xF2, 0x72, 0xFA, 0xD3, 0x63 };
+
+        // 256 00-01-00-00-00-20-00-00-00-0C-00-00-00-10-00-00-00-10-E7-DC-CE-66-DF-85-5A-32-3A-6B-B7-BD-7A-59-BE-45
+        private static readonly byte[] AES_256_GCM_Header = new byte[] { 0x00, 0x01, 0x00, 0x00, 0x00, 0x20, 0x00, 0x00, 0x00, 0x0C, 0x00, 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x10, 0xE7, 0xDC, 0xCE, 0x66, 0xDF, 0x85, 0x5A, 0x32, 0x3A, 0x6B, 0xB7, 0xBD, 0x7A, 0x59, 0xBE, 0x45 };
+
+        private static readonly Func<byte[], HashAlgorithm> _kdkPrfFactory = key => new HMACSHA512(key); // currently hardcoded to SHA512
+
+        private readonly byte[] _contextHeader;
+
+        private readonly Secret _keyDerivationKey;
+        private readonly int _derivedkeySizeInBytes;
+        private readonly IManagedGenRandom _genRandom;
+
+        public AesGcmAuthenticatedEncryptor(Secret keyDerivationKey, int derivedKeySizeInBytes, IManagedGenRandom? genRandom = null)
+        {
+            _keyDerivationKey = keyDerivationKey;
+            _derivedkeySizeInBytes = derivedKeySizeInBytes;
+
+            switch (_derivedkeySizeInBytes)
+            {
+                case 16:
+                    _contextHeader = AES_128_GCM_Header;
+                    break;
+                case 24:
+                    _contextHeader = AES_192_GCM_Header;
+                    break;
+                case 32:
+                    _contextHeader = AES_256_GCM_Header;
+                    break;
+                default:
+                    throw Error.CryptCommon_GenericError(); // should never happen
```

Comment:
Prefer `throw CryptoUtil.Fail("some message");`. It triggers a proper assert when under a debugger.